### ISSUE TITLE
Support case-sensitive file matching on Linux

### ISSUE THREAD

**Issue by chrisd8088** [github/VFSForGit#29]
As [noted](https://github.com/microsoft/VFSForGit/pull/1125#discussion_r289207160) in microsoft/VFSForGit#1125, we will need to refactor a [number](https://github.com/microsoft/VFSForGit/blob/ed07bc4deb5b35723d3780cab08a69694599af4f/GVFS/GVFS.Virtualization/FileSystem/FileSystemVirtualizer.cs#L153) of [points](https://github.com/microsoft/VFSForGit/blob/ed07bc4deb5b35723d3780cab08a69694599af4f/GVFS/GVFS.Platform.Mac/MacFileSystemVirtualizer.cs#L467) in the existing GVFS provider code to support case-sensitive file names on Linux.

**Comment by kivikakk** [comment]
These have been addressed in #1232. Note this does not imply compatibility with repositories with multiple files whose names differ only in case -- that is a separate (and quite larger) item.

**Comment by chrisd8088** [comment]
Unfortunately, certain basic operations fail on Linux with the existing code, e.g. (assuming `master` contains `foo.txt`):
```
git checkout -b foo
git mv foo.txt Foo.txt
git add .
git commit -m foo
git checkout master
```
leaves both `foo.txt` and `Foo.txt` in the working directory; the latter should obviously have been deleted by the second `checkout`.

This is due to case-insensitivity in the [modified paths database](https://github.com/microsoft/VFSForGit/blob/3f322d3a44375d2108e1b70ee5329c6bfc565ec7/GVFS/GVFS.Common/ModifiedPathsDatabase.cs#L21) and [virtualization code](https://github.com/microsoft/VFSForGit/blob/3f322d3a44375d2108e1b70ee5329c6bfc565ec7/GVFS/GVFS.Virtualization/Projection/GitIndexProjection.SortedFolderEntries.cs#L180), among other things, which results in case-differing-only filenames being dropped from the data sent back from the GVFS provider, via the [`virtual-filesystem` hook](https://github.com/microsoft/git/blob/9b1adf3543bd5831c925fba1c4a6078a83118dfd/read-cache.c#L1887), to `git` itself after it reads its `index` file.

Therefore, we will almost certainly need to ensure these internal data stores in the GVFS provider are case-sensitive when running on Linux.  Doing so (e.g., with the the WIP in 4fd47bea5bfb8eb91aba0b2406f1c41657145a66) appears to resolve the failure of the [`GitCommandsTests.CaseOnlyRenameFileAndChangeBranches()`](https://github.com/microsoft/VFSForGit/blob/3f322d3a44375d2108e1b70ee5329c6bfc565ec7/GVFS/GVFS.FunctionalTests/Tests/GitCommands/GitCommandsTests.cs#L398) functional test, which performs the `git` steps shown above.
